### DISCUSSION TITLE
Logger component in LWC - need more info on implementation

### DISCUSSION THREAD

**Discussion by dbrennan-008** [jongpie/NebulaLogger#450] [Q&A]
I am fairly new to LWC and I am trying to use nebula logger in our LWC components.  
Installed unlocked package in our sandbox dev env.
When I try to run jest test --> node ./node_modules/@salesforce/sfdx-lwc-jest/bin/sfdx-lwc-jest contactCreate.test.js
I am getting the following error
Cannot find module 'c/logger' from  'contact-app/main/default/lwc/contactCreate/contactCreate.html'

I have included the following at the top of my template markup
```
<template>
    <c-logger></c-logger>
   ...
   ...
   ...
</template>
```

And I am trying to use the following in .js 
```
const logger = this.template.querySelector("c-logger");
logger.error("Error creating contact");
```
I am sure I am missing something really basic here, do I need to include something in my package.json? or something? can someone please help..

**Comment by jamessimone** [comment]
@dbrennan008 when you use packaged dependencies - like Nebula's `<c-logger>` component - you need to explicitly tell Jest how to mock it:

```js
import Logger from 'c/logger';

jest.mock(
  'c/logger',
  () => ({
    error: jest.fn(),
    saveLog: jest.()
  }),
  { virtual: true }
);
```

Or something to that effect; you would then be able to say assert on the mocked version of `Logger` when testing the code-path where you're calling `logger.error`:

```js
expect(Logger.error).toHaveBeenCalled();
// additional asserts about what's been passed to the error function, plus that saveLog was called
```

**Reply by jamessimone** [reply]
Note that I haven't compiled this code and it may be easier for you to follow [one of the official stub examples](https://github.com/salesforce/sfdx-lwc-jest/blob/master/src/lightning-stubs/accordion/accordion.js) (which explicitly extend `LightningElement`). If you do go this route, make sure to also see how you wire `jest.config.js` up [to point to your explicit stubs](https://github.com/trailheadapps/lwc-recipes/blob/main/jest.config.js#L6).

**Reply by dbrennan-008** [reply]
Ah ok, that makes sense. Will try that out. 
It will be quite helpful for folks like me if we can have tests for LWCDemo.
Thank you.

**Comment by dbrennan-008** [comment]
Here is a sample simple test and I am getting Assert Violation: c() 2nd argument Ctor must be a function.  @jamessimone 

import { createElement } from "lwc";
import My_Component from "c/My_Component";
import nLogger from "c/logger";

jest.mock(
    'c/logger',
    () => ({
      error: jest.fn(),
      saveLog: jest.fn()
    }),
    { virtual: true }
);


describe('c-my-component', () => {
    
  it('should log error and save log', () => {
    // Instantiate the component
    const element = createElement("c-my-component", {
        is: My_Component
    });
    document.body.appendChild(element);

    // Call the logError method
    element.logError('Some error', 'Some details');

    // Verify that c-logger.error was called with the correct arguments
    expect(nLogger.error).toHaveBeenCalledWith('Some error', 'Some details');

    // Verify that c-logger.saveLog was called
    expect(nLogger.saveLog).toHaveBeenCalled();
  });
});
